core/src/main/scala/cats/derived/invariant.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2015 Miles Sabin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats
+package derived
+
+import shapeless._
+import util.VersionSpecific.OrElse
+
+import scala.annotation.implicitNotFound
+
+@implicitNotFound("Could not derive an instance of Invariant[${F}]")
+trait MkInvariant[F[_]] extends Invariant[F] {
+  def safeImap[A, B](fa: F[A])(g: A => Eval[B])(f: B => Eval[A]): Eval[F[B]]
+  def imap[A, B](fa: F[A])(g: A => B)(f: B => A): F[B] =
+    safeImap(fa)(g andThen Eval.now)(f andThen Eval.now).value
```

Comment:
I wonder what's the difference between using `f andThen Eval.now` and `a => Eval.later(f(a))`.
But I would say let's use the latter since it would be the same as `MkFunctor`.

Reply:
Ah good point, I think we would need the latter. Thanks.